### DISCUSSION TITLE
Option to solve AC simulation per frequency instead of using sweep

### DISCUSSION THREAD

**Discussion by hahschaa** [SpiceSharp/SpiceSharp#230] [Ideas]
Hi,

I really like this library. I have been using the AC simulation quite extensively for quite some time now, so first of all thanks for that :)

However, the way of retrieving data from the AC simulation using a sweep combined with `ExportSimulationData` is not the most convenient for me. That is because I usually want to get the results of 3 simulations simultaneously (but synced, so first the  3 results of frequency x, then the 3 results of frequency y, etc), and want to end the simulation prematurely if I found what I am looking for.

I have found ways around this; I can create a sweep that only yields the new value when I gathered the results of the three simulations, and I can use something like an `AutoResetEvent` inside of the `ExportSimulationData` to let the three simulations wait for each other.

While that works, if left me wondering why there is not an option to just do an AC simulation for one frequency at a time, so that I could do something like

```csharp
AC ac1 = new(..), ac2 = new(..), ac3 = new(..);

foreach (var f in [100, 1000, 10_000])
{
    var result1 = ac1.Simulate(f);
    var result2 = ac2.Simulate(f);
    var result3 = ac3.Simulate(f);
    if (...) break;
}
```

Something like this would make my code much cleaner, but I am wondering what your opinion is about this. I would gladly try to implement this I you are okay with that. Or maybe you have some other suggestions? Maybe I should've use parallel circuits in my simulations? Although I have the feeling that that still wouldn't make the code as clean as the example above as I still need the sweep/event combo. 

So, what are your thoughts? :)

Thank you in advance.

**Comment by svenboulanger** [comment]
Hi!

Simulating a single frequency should be fairly straightforward, since the constructor of `AC()` simply asks for an enumerable of frequency points. The [`Frequencies`](https://github.com/SpiceSharp/SpiceSharp/blob/d3565b0152ac50af8d9e215619eb9c1e96c86252/SpiceSharp/Simulations/Base/Frequency/FrequencyParameters.cs#L31) parameter can also be set directly. So doing

```csharp
AC ac1 = new("AC1", new[] { my_freq });
// or
ac1.FrequencyParameters.Frequencies = new[] { my_freq };
```

should work. If your circuit doesn't change, then you can also use the `Rerun()` method to avoid recreating behaviors every time. However, it will rerun an operating point analysis and temperature-dependent calculation giving you some overhead between runs.

Also note that the event is "blocking" for the simulation, so it may be possible to use multithreading magic to wait/sync the three simulations.

**Comment by hahschaa** [comment]
Thanks for your reply Sven,

That is indeed what I do now; spin up three threads each containing a simulation and pump the results into an `IProducerConsumerCollection<T>` so that the parent thread can consume them as if the simulations are running at the same pace. I indeed use the blocking behaviour of the event for that; the simulations can only continue if the results of all three simulations are in.

So I already have found my way around it. I was just wondering if you'd see any merit in an additional way of consuming the results, alongside the existing way. It feels like I am bending myself into corners needlessly here, so I was wondering if I could do something about it.

I will take a look at the `Rerun()` approach. Like you said I didn't want to create new simulation for each frequency because I expected there to be some overhead in that. Not that there is no overhead in syncing the threads of course.

**Comment by hahschaa** [comment]
I created a small proof of concept of the way I suggested it:
https://github.com/SpiceSharp/SpiceSharp/compare/development...hahschaa:SpiceSharp:feature/ac-iterator?diff=split&w=

Is this something you think has potential after some refinement? Or do you think it’s not suitable for inclusion?
The added unit test shows how the POC could be used:

```csharp
 // Create the circuit
var voltage = 1d;
var ckt = new Circuit(
    new VoltageSource("V1", "in", "0", 0).SetParameter("acmag", voltage),
    new Resistor("R1", "in", "out", 10),
    new Capacitor("C1", "out", "0", 20e-6)
);

// Create the ac analysis
var ac = new AC("ac 1");
var export = new ComplexPropertyExport(ac, "R1", "i");
var iterator = ac.CreateIterator(ckt); // <-- Create the object that allows for single frequencies to be passed

var frequencies = new double[] { 1000, 5000, 10_000, 50_000 };
var expectedImpedance = new[] { 12.7799, 10.126, 10.0326, 10.0013 };

for (var i = 0; i < frequencies.Length; i++)
{
    double f = frequencies[i];
    var result = iterator.ExecuteSingle(f); // <-- Run the single frequency
    Assert.That(result.Frequency, Is.EqualTo(f).Within(0.1).Percent);
    var current = export.Value.Magnitude;
    var impedance = voltage / current;
    Assert.That(impedance, Is.EqualTo(expectedImpedance[i]).Within(0.1).Percent);
}
```

**Reply by svenboulanger** [reply]
To be honest, I really like the idea, but I'm inclined to refactor the `Run` method such that they return an `IEnumerable<>`. You would then be able to do something like:

```csharp
var sim = new Transient("tran", 1e-9, 1e-6);
var v = new RealVoltageExport(sim, "a");

// Automatically starts a transient simulation from start to finish
foreach (var point in sim.Run(ckt))
    Console.WriteLine(v.Value);
```

You can then also use it like you intend it:

```csharp
var sim = new Transient("tran", 1e-6);
var v = new RealVoltageExport(sim, "a");
var it = sim.Run(ckt).GetEnumerator();

while (it.MoveNext())
    Console.WriteLine(v.Value);
```

However, there are a few questions then:

- Do we want to retain backward compatibility? In that case, we might need a `RunIterable`/`RerunIterable` methods, and have `Run`/`Rerun` simply wrap those methods with a `foreach` statement.
- What is the value returned by the iterator? I'd like this value to be agnostic to the type of simulation. Perhaps we can return a type of `enum` that tells at which "point" the simulator is. This might make the whole event-based system obsolete, or also part of a refactor.

```csharp
foreach (var state in sim.Run(ckt))
{
    if (state == SimulationState.ExportData)
        Console.WriteLine(v.Value);
    else if (state == SimulationState.AfterUnsetup)
        Console.WriteLine("Finished")
}
```

This feels like a pretty big change in the API, so it'd probably be a major version increase.

**Reply by svenboulanger** [reply]
I've been playing around with the idea, but it does seem not very straightforward as it is touching a lot of different things.

I would like to avoid having both events _and_ an iterator as IMO it will lead to a very chaotic code architecture (and it is already relatively complex) and I'd like not to be too stuck on backward compatibility. There are quite a few parts that are using events, which I'd like to refactor out first.

I do believe that your proposal is long-term more sustainable and user-friendly. I will continue to investigate.

**Reply by hahschaa** [reply]
There is still a small difference, because what I proposed would allow this:

```csharp
var iteratorA = ac1.CreateIterator(ckt1);
var iteratorA = ac2.CreateIterator(ckt2);
for (var i = 0; i < frequencies.Length; i++)
{
    double f = frequencies[i];
    var r1 = iteratorA.ExecuteSingle(f);
    var r2 = iteratorB.ExecuteSingle(f);
}
```

While with your proposal that would become:
```csharp
var iteratorA = ac1.Run(ckt1).GetEnumerator();
var iteratorB = ac2.Run(ckt2).GetEnumerator();
while (iteratorA.MoveNext())
{
    if (!iteratorB.MoveNext())
        throw new InvalidOperationException();
    Console.WriteLine(v.Value);
}
```

Syncing two sims has becomes miles easier indeed, but the checking if two simulators are indeed in sync is still less readable imho.

But anyway, I'm glad that you like the idea and would gladly help to bring either solutions to fruition. 

Removing the events completely would indeed be a major breaking change, and I get your point that it might lead to chaotic code, but from a outside perspective it might be desirable? Maybe other projects that use this library have a good reason to want things to stay the way they are? 

I could try to simplify the events a bit (if possible, I havn't looked). I think events only have to be observed from the outside, not internally within the simulations. Maybe we can see from there?

**Reply by hahschaa** [reply]
Oh, should've taken a look before posting. I see that the events are not observed within the simulations themselves (except for `AfterLoad` in `BiasingSimulation`). I thought that was what you meant by "There are quite a few parts that are using events, which I'd like to refactor out first.". So what exactly did you mean by that?

**Reply by svenboulanger** [reply]
Basically, the iterator would not only stop when exporting data, but also at all locations where there is currently an event in place. The user can then simply execute code on the simulation. In the meantime I have been experimenting and discovering some cons with that approach, mainly for tight loops such as for the `BeforeLoad` and `AfterLoad` events which are called very often. Perhaps the iterator should only stop when exporting data.

But you're right, I partially misunderstood your approach. Nevertheless, I think I'd like to stick to an enumerable from the `Run` method, as it allows a neat way to pause simulations, synchronize simulations, etc. The vision I have of your code would look something like this:

```csharp
var iterator1 = ac1.Run(ckt1).GetEnumerator();
var iterator2 = ac2.Run(ckt2).GetEnumerator();
for (var i = 0; i < frequencies.Length; i++)
{
    ac1.NextFrequency = ac2.NextFrequency = frequencies[i];
    ac1.MoveNext();
    ac2.MoveNext();
}
```

Then it would use more of the "standard" C# enumerator approach.

**Reply by hahschaa** [reply]
Ah, that indeed looks very nice and clean :)

If I understand you correctly now, the `MoveNext()` would also return on `BeforeLoad`, `AfterLoad` and the rest of the events?
Maybe in that case it could be useful to pass an enum or something like that to specify on which steps you'd want the simulation to yield a result. And then the return type could be som kind of abstact result type like:

```csharp
public abstract record SimResult(...)
{
    public sealed record BeforeLoadResult(...) : SimResult(...)
    {
        ...
    }

    public sealed record ExportResult(...) : SimResult(...)
    {
        ...
    } 

   // etc
}

I don't know if that would make sense.

**Reply by svenboulanger** [reply]
### On the enumerable

> If I understand you correctly now, the MoveNext() would also return on `BeforeLoad`, `AfterLoad` and the rest of the events?

Yes, that was the idea that your approach gave me. However, after fiddling with it I noticed that it might not be a good idea to "yield" inside very tight loops. The `BeforeLoad` and `AfterLoad` are in that regard inside the tightest loops.
The issue I see with `yield return` in this scenario is that I am nesting methods, each one returning an `IEnumerable<...>`. To demonstrate, something akin to:

```csharp
// Inside Simulation.cs (the abstract description)
public IEnumerable<SimulationStatus> Run(IEntityCollection ckt)
{
    foreach (var status in Setup(ckt))
    { yield return status; }

    foreach (var status in Validate())
    { yield return status; }

    foreach (var status in Execute())
    { yield return status; }
}

// Inside the implementation, for example the Execute()
public override IEnumerable<SimulationStatus> Execute()
{
    while (main_loop)
    {
        foreach (var status in Iterate())
        { yield return status; }

        // Export here
        yield return Constant.ExportMe;
    }
}

private IEnumerable<SimulationStatus> Iterate()
{
    bool notConverging = true;
    while (notConverging)
    {
        yield return Constant.BeforeLoad;
        Load();
        yield return Constant.AfterLoad;

        // Solve
        Solve();
        notConverging = Convergent();
    }
}
```

The `Constant.BeforeLoad` and `Constant.AfterLoad` yield return statements are in tight loops, and have to "bubble up" a few levels introducing some overhead. In most cases, these yields aren't even used... Hence I'm considering using events only for performance-critical loops, and for example exports or top-level "break points" to be yielding the simulation status.

> Maybe in that case it could be useful to pass an enum or something like that to specify on which steps you'd want the simulation to yield a result.

I considered using an enum first, but this makes it difficult to deal with for third-party libraries. A class, struct or interface would probably be suited better, but when I was considering you would like to check efficiently what the current simulation status is, I thought maybe just a unique integer may be fine, defined as a constant by the simulation itself.

For example, the following code would not be possible with structs since they cannot be stored in a `const` variable:

```csharp
foreach (int status in ac.Run(ckt))
{
    switch (status)
    {
        case SimulationStatus.ExportOperatingPoint:
            // Store OP point (real data)
            break;

        case SimulationStatus.ExportSmallSignal:
            // Store AC point data (complex data)
            break;
    }
}
```

### Choosing a next frequency point

I've also found a way for what you want to do that doesn't need a `NextFrequency` property on an AC simulation in order to achieve the effect that you want. You can make a "custom" enumerable class that simply keeps yielding a new value as long as it receives a new value. I've written a small demo of that [on .NET Fiddle](https://dotnetfiddle.net/cJBHdh).

So the result would be, using the `DynamicEnumerable<T>` class:

```csharp
var de = new DynamicEnumerable<double>();
var ac1 = new AC("ac 1", de);
var ac2 = new AC("ac 2", de);

var it1 = ac1.Run(ckt1).GetEnumerator();
var it2 = ac2.Run(ckt2).GetEnumerator();

for (int i = 0; i < frequencies.Length; i++)
{
    de.Next = frequencies[i];
    it1.MoveNext();
    it2.MoveNext(); // Although you might want to iterate until the first relevant export point
}
```

**Reply by hahschaa** [reply]
The enum that I proposed was only meant to control which type of statuses that would be yielded. I don't think that would be hard to use for third parties. Something like this:

```csharp
[Flags]
public enum ESimulationStatusType
{
    Setup,
    Validation,
    Execution,
}
``` 

That you can then pass to the `Run` method, something like:
```csharp
public IEnumerable<SimulationStatus> Run(
    IEntityCollection ckt,
    ESimulationStatusType statuses = ESimulationStatusType.Execution)
{
    var suppressSetupStatus = !statuses.HasFlag(ESimulationStatusType.Setup);
    foreach (var status in Setup(ckt, suppressSetupStatus))
        { yield return status; }

    // etc.
}

public IEnumerable<SimulationStatus> Setup(IEntityCollection ckt, bool suppressStatus)
{
    if (!suppressStatus)
        yield new SimulationStatus(...);
}
```

But I have no idea how performant that would be. Could be better with something like [FastEnum](https://github.com/xin9le/FastEnum), but I don't know.

I also have a feeling that somehow it should be possible to push a collection down instead of letting everything bubble up. Something like a `IProducerConsumerCollection<T>`. That way each layer can directly add statuses to the result, but I guess the overhead of that would also be non-negligible. And pushing a "regular" delegate down would of course just be the same as an event, so that also doesn't make much sense..


About the `SimulationStatus` implementation I think we are talking about the same thing (I called them `SimResult` in my previous post). I indeed meant that you could switch on these types like you mentioned. So combined with the enum that would be something like:
```csharp
foreach (int status in ac.Run(ckt, ESimulationStatusType.Setup | ESimulationStatusType.Validation))
{
    switch (status)
    {
        case SimulationStatus.Setup:
            // No idea what to do here
            break;

        case SimulationStatus.ExportOperatingPoint:
            // Store OP point (real data)
            break;

        case SimulationStatus.ExportSmallSignal:
            // Store AC point data (complex data)
            break;
    }
}
```

**Reply by hahschaa** [reply]
Btw, how likely is it that someone is interested in the validation status while already iterating? Because you can just perform a validation before starting iterating, and in the case that you want to change the circuit during the loop it would also be possible to do a validation whenever the circuit is changed. But maybe this is just a feature that I don't use..

**Reply by svenboulanger** [reply]
> The enum that I proposed was only meant to control which type of statuses that would be yielded. I don't think that would be hard to use for third parties.

The issue I have is that simulations may wish to add their enum values, but this cannot be done by external libraries. Let's say we define the enum values for export types `ExportOP`, `ExportAC` and `ExportTransient`.
Now an external library wants to make a noise simulation, and wants to add a new export status value called `ExportNoise`. Enums are too rigid for that...

[edit] Hence, sticking to an int should work.

Having a kind of filter to pass with the `Run(...)` method does make sense to me, but I'd definitely want to keep it as light-weight as possible.

**Comment by svenboulanger** [comment]
I have created a pull request to show what I would like to add to Spice#.

[https://github.com/SpiceSharp/SpiceSharp/pull/236](https://github.com/SpiceSharp/SpiceSharp/pull/236)

**Comment by hahschaa** [comment]
I just returned from my holiday, very cool to see you've worked on this. I'll take a look as soon as possible :)

**Comment by hahschaa** [comment]
I have been trying your branch and I really like what you've done :)
I quickly threw together the helper class below, which makes my use case quite a lot simpler.

```csharp
public class AcSyncRunner : IDisposable
{
    private readonly FakeSweep frequencyControl = new();
    private readonly Dictionary<Circuit, SimData> simulations;

    public AcSyncRunner(IReadOnlyCollection<Circuit> circuits)
    {
        this.simulations = new(circuits.Count);

        foreach (var circuit in circuits)
        {
            var ac = new AC(string.Empty, this.frequencyControl);
            this.simulations[circuit] = new(circuit, ac);
        }
    }

    public IReadOnlyDictionary<Circuit, SimData> SimulationData => this.simulations;

    public void Next(double frequency)
    {
        this.frequencyControl.Next = frequency;
        foreach (var simulation in this.simulations.Values)
        {
            simulation.Runner.MoveNext();
            Debug.Assert(simulation.Simulation.Frequency - frequency < 1);
        }
    }

    public void Dispose()
    {
        this.frequencyControl.Next = null;
        foreach (var sim in simulations.Values)
        {
            while (sim.Runner.MoveNext())
            {
                // Discard.
            }

            Debug.Assert(sim.Simulation.Status != SimulationStatus.Running);
            sim.Runner.Dispose();
        }
    }

    public sealed record SimData(Circuit Circuit, AC Simulation) : IDisposable
    {
        private readonly Dictionary<string, ComplexPropertyExport> current = [];
        private readonly Dictionary<string, ComplexPropertyExport> voltage = [];

        private IEnumerator<int>? runner;

        public IEnumerator<int> Runner =>
            this.runner ??= this.Simulation
                .Run(this.Circuit, AC.ExportSmallSignal)
                .GetEnumerator();

        public Complex Current(string refdes)
        {
            ref var exp = ref CollectionsMarshal.GetValueRefOrAddDefault(this.current, refdes, out var exist);
            if (!exist) exp = new(this.Simulation, refdes, "i");
            return exp.Value;
        }

        public Complex Voltage(string refdes)
        {
            ref var exp = ref CollectionsMarshal.GetValueRefOrAddDefault(this.voltage, refdes, out var exist);
            if (!exist) exp = new(this.Simulation, refdes, "v");
            return exp.Value;
        }

        public void Dispose() => runner?.Dispose();
    }

    private sealed record FakeSweep : IEnumerable<double>
    {
        public double? Next { get; set; } = null;

        public IEnumerator<double> GetEnumerator()
        {
            while (this.Next is { } n)
                yield return n;
        }

        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }
}
```

I can use it like so:
```csharp
var ckt1 = new Circuit(...), ckt2 = new Circuit(...), ckt3 = new Circuit(...);
using var runner = new AcSyncRunner([ckt1, ckt2, ckt3]);
foreach (var f in [10, 100, 1000])
{
    runner.Next(f);
    var r1V = runner.SimulationData[ckt1].Voltage("R1");
    var r1C = runner.SimulationData[ckt1].Current("R1");
    ...
    if (...) break;
}
```

That already looks so much more clean than what I have with the events.
